compiler/qsc_eval/src/lib.rs

```diff
@@ -154,26 +139,55 @@ pub fn eval_expr(
     out: &mut impl Receiver,
 ) -> Result<Value, (Error, Vec<Frame>)> {
     state.push_expr(expr);
-    state.resume(globals, env, sim, out, &[])?;
-    Ok(state.pop_val())
-}
-
-pub fn eval_push_expr(state: &mut State, expr: ExprId) {
-    state.push_expr(expr);
+    let res = state.eval(globals, env, sim, out, &[], &StepAction::Continue)?;
+    let StepResult::Return(value) = res else {
+        panic!("eval_expr should always return a value");
+    };
+    Ok(value)
 }
 
-/// Continues evaluation given current state with the given context.
+/// Evaluates the given stmt with the given context.
 /// # Errors
 /// Returns the first error encountered during execution.
-pub fn eval_resume(
-    state: &mut State,
+/// # Panics
+/// On internal error where no result is returned.
+pub fn eval_stmt(
+    stmt: StmtId,
     globals: &impl NodeLookup,
     env: &mut Env,
     sim: &mut impl Backend<ResultType = impl Into<val::Result>>,
-    out: &mut impl Receiver,
-    breakpoints: &[StmtId],
-) -> Result<Option<StmtId>, (Error, Vec<Frame>)> {
-    state.resume(globals, env, sim, out, breakpoints)
+    package: PackageId,
+    receiver: &mut impl Receiver,
+) -> Result<Value, (Error, Vec<Frame>)> {
+    let mut state = State::new(package);
+    state.push_stmt(stmt);
+    let res = state.eval(globals, env, sim, receiver, &[], &StepAction::Continue)?;
+    let StepResult::Return(value) = res else {
+        panic!("eval_stmt should always return a value");
+    };
+    Ok(value)
+}
+
+/// The type of step action to take during evaluation
+pub enum StepAction {
```

Comment:
This could have a derive to make other code simpler:

```suggestion
/// The type of step action to take during evaluation
#[derive(PartialEq)]
pub enum StepAction {
```